internal/cmd/agent/controllers/bundledeployment/controller.go

```diff
@@ -130,11 +136,110 @@ func (h *handler) DeployBundle(bd *fleet.BundleDeployment, status fleet.BundleDe
 	status.Release = release
 	status.AppliedDeploymentID = bd.Spec.DeploymentID
 
+	err = h.setNamespaceLabelsAndAnnotations(bd, release)
+
+	if err != nil {
+		return fleet.BundleDeploymentStatus{}, err
+	}
+
 	// Setting the error to nil clears any existing error
 	condition.Cond(fleet.BundleDeploymentConditionInstalled).SetError(&status, "", nil)
 	return status, nil
 }
 
+func (h *handler) setNamespaceLabelsAndAnnotations(bd *fleet.BundleDeployment, releaseID string) error {
+	if bd.Spec.Options.NamespaceLabels == nil && bd.Spec.Options.NamespaceAnnotations == nil {
+		return nil
+	}
+
+	ns, err := h.fetchNamespace(releaseID)
+	if err != nil {
+		return err
+	}
+
+	if reflect.DeepEqual(bd.Spec.Options.NamespaceLabels, ns.Labels) && reflect.DeepEqual(bd.Spec.Options.NamespaceAnnotations, ns.Annotations) {
+		return nil
+	}
+
+	if ns.Annotations == nil && bd.Spec.Options.NamespaceAnnotations != nil {
+		ns.Annotations = map[string]string{}
+	}
+	if bd.Spec.Options.NamespaceLabels != nil {
+		addLabelsFromOptions(ns.Labels, *bd.Spec.Options.NamespaceLabels)
+	}
+	if bd.Spec.Options.NamespaceAnnotations != nil {
+		addAnnotationsFromOptions(ns.Annotations, *bd.Spec.Options.NamespaceAnnotations)
+	}
+	err = h.updateNamespace(ns)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (h *handler) updateNamespace(ns *corev1.Namespace) error {
+	u, err := runtime.DefaultUnstructuredConverter.ToUnstructured(ns)
+	if err != nil {
+		return err
+	}
+	_, err = h.dynamic.Resource(nsResource).Update(h.ctx, &unstructured.Unstructured{Object: u}, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (h *handler) fetchNamespace(releaseID string) (*corev1.Namespace, error) {
+	// releaseID is composed of release.Namespace/release.Name/release.Version
+	namespace := strings.Split(releaseID, "/")[0]
+	list, err := h.dynamic.Resource(nsResource).List(h.ctx, metav1.ListOptions{
+		LabelSelector: "name=" + namespace,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(list.Items) == 0 {
+		return nil, errors.New("namespace not found")
+	}
+	var ns corev1.Namespace
+	err = runtime.DefaultUnstructuredConverter.
+		FromUnstructured(list.Items[0].Object, &ns)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ns, nil
+}
+
+func addLabelsFromOptions(nsLabels map[string]string, optLabels map[string]string) {
+	for k, v := range optLabels {
+		nsLabels[k] = v
+	}
+
+	// Delete labels not defined in the options.
+	// Keep the name label as it is added by helm when creating the namespace.
+	for k := range nsLabels {
+		if _, ok := optLabels[k]; k != "name" && !ok {
+			delete(nsLabels, k)
+		}
+	}
+}
+
+func addAnnotationsFromOptions(nsAnnotations map[string]string, optAnnotations map[string]string) {
```

Comment:
```suggestion
// addAnnotationsFromOptions updates nsAnnotations so that it only contains all annotations specified in optAnnotations.
func addAnnotationsFromOptions(nsAnnotations map[string]string, optAnnotations map[string]string) {
```